src/librustc_mir/interpret/eval_context.rs

```diff
@@ -838,7 +855,13 @@ impl<'a, 'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> EvalContext<'a, 'mir, 'tcx, M
 
         match layout.variants {
             layout::Variants::Single { index } => {
-                return Ok(index as u128);
+                if layout.for_variant(&self, index).abi == layout::Abi::Uninhabited {
+                    return Ok(index as u128);
+                }
+                let discr_val = ty.ty_adt_def().map_or(
+                    index as u128,
+                    |def| def.discriminant_for_variant(*self.tcx, index).val);
+                return Ok(discr_val);
```

Comment:
I think `discriminant_for_variant` is the right value. cc @oli-obk did you ever fix the tagged case, to convert the tag to the discriminant? (shouldn't affect this PR unless there are other hacks elsewhere)